components/authz-service/server/v2/project_update_manager.go

```diff
@@ -390,7 +415,82 @@ func (m *CerealProjectUpdateManager) Status() (ProjectUpdateStatus, error) {
 	return projectUpdateInstance, nil
 }
 
-func (w *workflowInstance) collectAllDomainServiceJobStatuses() []project_update_tags.JobStatus {
+func (w *workflowInstance) collectProgress() (float32, error) {
+	if !w.IsRunning() {
+		return 1, nil
+	}
+	domainServicesUpdateInstance, err := w.GetUpdateDomainServicesInstance()
+	if err != nil {
+		return 0, err
+	}
+
+	progress := []float32{}
+	weight := []int{}
+
+	for _, d := range domainServicesUpdateInstance.ListSubWorkflows() {
+		subWorkflow, err := domainServicesUpdateInstance.GetSubWorkflow(d)
+		if err != nil {
+			logrus.WithError(err).Errorf("failed to get subworkflow for %q", d)
+			continue
+		}
+
+		if d == serialSubworkflowName {
+			state := project_update_tags.SerializedProjectUpdateWorkflowState{}
+			if subWorkflow.IsRunning() {
+				if err := subWorkflow.GetPayload(&state); err != nil {
+					logrus.WithError(err).Errorf("failed to get payload for %q", d)
+					continue
+				}
+			} else {
+				if err := subWorkflow.GetResult(&state); err != nil {
+					logrus.WithError(err).Errorf("failed to get result for %q", d)
+					continue
+				}
+			}
+
+			if state.Phase == project_update_tags.SerializedProjectUpdateWorkflowPhaseRunning && state.Running.TotalTasks > 0 {
+				progress = append(progress,
+					state.Running.RunningTask.LastStatus.PercentageComplete)
+				weight = append(weight, 1)
+			}
+			progress = append(progress, 1)
+			weight = append(weight, state.Running.CompletedTasks)
+
+			progress = append(progress, 0)
+			weight = append(weight, len(state.Running.RemainingTasks))
+		} else {
+			payload := project_update_tags.DomainProjectUpdateWorkflowPayload{}
+			if subWorkflow.IsRunning() {
+				if err := subWorkflow.GetPayload(&payload); err != nil {
+					logrus.WithError(err).Errorf("failed to get payload for %q", d)
+					continue
+				}
+				progress = append(progress, payload.MergedJobStatus.PercentageComplete)
+				weight = append(weight, 1)
+			} else {
+				progress = append(progress, 1)
+				weight = append(weight, 1)
+			}
+		}
+	}
+
+	totalWeight := 0
+	for i := range weight {
+		totalWeight += weight[i]
+	}
+
+	if totalWeight == 0 {
+		return 0, nil
+	}
+	var prog float32
+	for i := range weight {
+		prog += (progress[i] * float32(weight[i]) / float32(totalWeight))
+	}
+
+	return prog, nil
+}
```

Comment:
I am not exactly sure how this is working, but I put in example numbers and I got a good estimate of completion. I have also run this many times and the percentage looks correct.

Reply:
Ignoring nodemanager:
The general idea is each index is a task. We give each task the same weight, 1.
So the total weight would be the number of indexes.
To get there:
The current running task has a weight of 1, and its progress is whatever we get from looking at the ES task (some number less than 1).
Then you have the indexes that are complete: the progress of  these is 1 because they are all complete, but it has a weight of the number completed indexes
Then you have some number of indexes that are not running yet. These have a progress of 0, and a weight of the number of remaining indexes.

So: 
total weight = denominator =  1 (which is the current running index) + # completed index + # remaining indexes
numerator = progress of running task + 1 (progress of completed tasks which is 100%) * # completed tasks + 0 (which is the progress of the incomplete tasks) * # incomplete tasks

Reply:
Thank you for explaining it.